src/toast/scripts/benchmarking_utilities.py

```diff
@@ -128,29 +127,58 @@ def get_minimum_memory_use(args, n_nodes, n_procs, total_samples, full_pointing)
     # returns the group_nodes and n_detector that minimize memory usage
     return (group_nodes_best, n_detector_best, num_obs_best, memory_used_bytes_best)
 
+def maximize_nb_samples(args, n_nodes, n_procs, full_pointing, available_memory_bytes, per_process_overhead_bytes=1024 ** 3):
+    """
+    Finds the largest number of samples that can fit in the available memory.
+    Returns 1 if not number of sample fits in memory.
+    One can set `per_process_overhead_bytes` (which defaults to 1GB) to define a number of bytes 
+    that will be consummed by each process, independently of the number of samples.
+    """
+    # returns true if a number of samples can fit in memory
+    def fits_in_memory(nb_samples):
+        (_, _, _, memory_used_bytes) = get_minimum_memory_use(args, n_nodes, n_procs, nb_samples, full_pointing)
+        return memory_used_bytes + n_procs * per_process_overhead_bytes < available_memory_bytes
+    # finds an upper-bound on the number of samples that can fit in memory
+    max_samples = 2
+    while fits_in_memory(max_samples):
+        max_samples *= 2
+    min_samples = max_samples // 2
+    # finds the largest number of samples that *does* fit in memory
+    # using a binary search between min_samples (fits in memory)
+    # and max_samples (does not fit in memory)
+    mid_samples = (min_samples + max_samples) // 2
+    while mid_samples != min_samples:
+        if fits_in_memory(mid_samples):
+            min_samples = mid_samples
+        else:
+            max_samples = mid_samples
+        mid_samples = (min_samples + max_samples) // 2
+    return mid_samples
 
 def select_case(
-    args, n_procs, n_nodes, avail_node_bytes, full_pointing, world_comm, log
+    args, n_procs, n_nodes, avail_node_bytes, full_pointing, world_comm, log, per_process_overhead_bytes=1024 ** 3
```

Comment:
I find `2**30` a more intuitive way of writing 1GB but of course this is a matter of taste.

Reply:
I used `1024 ** 3` as its used in other part of the benchmark code inherited from the TOAST2 version (to convert bytes number when displaying).